proposals/NNNN-generalize-async-sequence.md

```diff
@@ -0,0 +1,311 @@
+
+Finally, `next()` always running on the generic executor is the source of unnecessary hops between an actor and the generic executor.
+
+## Proposed solution
+
+This proposal replaces the `@rethrows` attribute on `AsyncSequence` and `AsyncIteratorProtocol` with a primary `Failure` associated type, adds a new protocol requirement to `AsyncIteratorProtocol` that generalizes the existing `next()` requirement by throwing the `Failure` type, and adds an `isolated` parameter to the new requirement to abstract over actor isolation:
+
+```swift
+@available(SwiftStdlib 5.1, *)
+protocol AsyncIteratorProtocol<Element, Failure> {
+  associatedtype Element
+
+  mutating func next() async throws -> Element?
+
+  @available(SwiftStdlib 5.11, *)
+  associatedtype Failure: Error = any Error
+
+  @available(SwiftStdlib 5.11, *)
+  mutating func nextElement(_ actor: isolated (any Actor)?) async throws(Failure) -> Element?
+}
+
+@available(SwiftStdlib 5.1, *)
+public protocol AsyncSequence<Element, Failure> {
+  associatedtype AsyncIterator: AsyncIteratorProtocol
+  associatedtype Element where AsyncIterator.Element == Element
+
+  @available(SwiftStdlib 5.11, *)
+  associatedtype Failure = AsyncIterator.Failure where AsyncIterator.Failure == Failure
+
+  func makeAsyncIterator() -> AsyncIterator
+}
+```
+
+The new `nextElement()` has a default implementation so that conformances will continue to behave as they do today. Code generation for `for-in` loops will switch over to calling `nextElement()` instead of `next()` when the context has appropriate availability.
+
+## Detailed design
+
+### Adopting typed throws
+
+Concrete `AsyncSequence` and `AsyncIteratorProtocol` types determine whether calling `next()` can `throw`. This can be described in each protocol with a `Failure` associated type that is thrown by the `AsyncIteratorProtcol.nextElement()` requirement. Describing the thrown error with an associated type allows conformances to fulfill the requirement with a type parameter, which means that libraries do not need to expose separate throwing and non-throwing concrete types that otherwise have the same async iteration functionality.
+
+#### Error type inference from `for try await` loops
+
+The `Failure` associated type is only accessible at runtime in the Swift 5.11 standard library; code running against older standard library versions does not include the `Failure` requirement in the witness tables for `AsyncSequence` and `AsyncIteratorProtocol` conformances. This impacts error type inference from `for try await` loops.
```

Comment:
I think it would potentially make more sense to move this below a couple paragraphs, once we've introduced what the inference behavior actually is (i.e., as a caveat to the `MyAsyncIterator` example, and just before "When the throw error type of an `AsyncIteratorProtocol`..."

Reply:
Oh hmm, I meant this as a general statement that doesn't just apply to error type inference in `for try await` loops. Maybe I should move this up into the previous section to describe why the `Failure` associated type has Swift 5.11 availability. Then this section can say that because `Failure` is only available in Swift 5.11, thrown error type inference depends on availability context. Thoughts?

Reply:
Yes that reorganization makes sense to me!